util/pkg/vfs/vaultcontext.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vfs
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	vault "github.com/hashicorp/vault/api"
+)
+
+func newVaultClient(addr string) (*vault.Client, error) {
+
+	config := &vault.Config{
+		Address:    addr,
+		HttpClient: &http.Client{},
+	}
+	client, err := vault.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	err = authenticateVaultClient(client)
+	if err != nil {
+		return nil, err
+	}
+	return client, err
+}
+
+func authenticateVaultClient(client *vault.Client) error {
+	token := os.Getenv("VAULT_TOKEN")
+	if token != "" {
+		client.SetToken(token)
+	} else {
+		awsAuth(client)
+	}
+	return nil
+}
+
+func awsAuth(client *vault.Client) error {
+	config := aws.NewConfig()
+	config = config.WithCredentialsChainVerboseErrors(true)
+	session, err := session.NewSession(config)
+	if err != nil {
+		return err
+	}
+	svc := sts.New(session)
+	var params *sts.GetCallerIdentityInput
+	stsRequest, _ := svc.GetCallerIdentityRequest(params)
```

Comment:
Perhaps shove an `X-Vault-AWS-IAM-Server-ID` header into the request before signing it?

Reply:
That is why that code is there. But I think this needs to be configurable as it has to match the config on the vault side, and I wasn't entirely sure how to best do this. I could do this as another query string parameter, I guess.

Reply:
I would say pass the domain name of the Vault server.

(I really don't like how lax Vault is about authentication replay attacks.)

Reply:
Done